### PR TITLE
Improve API test for datatype change

### PR DESCRIPTION
**Pull request by simonbray**
As suggested by @mvdbeek here: https://github.com/galaxyproject/galaxy/pull/9374#issuecomment-696629916

### CODE REVIEW

File: lib/galaxy/managers/datasets.py [hunk 1]
```diff
@@ -691,7 +691,7 @@ def deserialize_datatype(self, item, key, val, **context):
         if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
             raise exceptions.RequestParameterInvalidException("The target datatype does not allow datatype changes.")
         if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
-            raise Exception("Dataset metadata could not be updated because it is used as input or output of a running job.")
+            raise exceptions.RequestParameterInvalidException("Dataset metadata could not be updated because it is used as input or output of a running job.")
```
Comment (simonbray) [inline lib/galaxy/managers/datasets.py:694]: I changed this back as suggested - not sure if you and @nsoranzo finished arguing it out though.

**Review by mvdbeek — APPROVED** [review]
Looks good, thanks @simonbray.

**Comment by nsoranzo** [comment]
I think there is another part of this test that is prone to fail, see this failure from another PR: https://jenkins.galaxyproject.org/job/docker-api-py3/5657/testReport/junit/lib.galaxy_test.api.test_datasets/DatasetsApiTestCase/test_update_datatype/

File: lib/galaxy_test/api/test_datasets.py [hunk 2]
```diff
@@ -143,8 +145,20 @@ def test_update_datatype(self):
...
+            "cat_data_and_sleep",
+            inputs,
+            self.history_id,
+            assert_ok=False,
+        )
+        queued_id = run_response.json()["outputs"][0]["id"]
+
         update_while_incomplete_response = self._put(  # try updating datatype before upload is complete
-            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=queued_id),
```
Comment (mvdbeek) [inline lib/galaxy_test/api/test_datasets.py:161]: And this shouldn't return a 500 status code anymore

**Comment by simonbray** [comment]
> I think there is another part of this test that is prone to fail, see this failure from another PR: https://jenkins.galaxyproject.org/job/docker-api-py3/5657/testReport/junit/lib.galaxy_test.api.test_datasets/DatasetsApiTestCase/test_update_datatype/

It turns out the `wait_for_history` was not doing anything as it only checks the state of the history, not the datasets contained.

File: lib/galaxy_test/api/test_datasets.py [hunk 3]
```diff
@@ -143,12 +145,24 @@ def test_update_datatype(self):
...
+        }
+        run_response = self.dataset_populator.run_tool(
+            "cat_data_and_sleep",
+            inputs,
+            self.history_id,
+            assert_ok=False,
+        )
+        queued_id = run_response.json()["outputs"][0]["id"]
+
         update_while_incomplete_response = self._put(  # try updating datatype before upload is complete
```
Comment (nsoranzo) [inline lib/galaxy_test/api/test_datasets.py]:
```suggestion
        update_while_incomplete_response = self._put(  # try updating datatype while used as output of a running job
```

**Pull request merged**